lib/snippet-body.pegjs

```diff
@@ -1,82 +1,141 @@
-    }, []);
+  function makeInteger(i) {
+    return parseInt(i.join(""), 10);
   }
 }
-bodyContent = content:(tabStop / bodyContentText)* { return content; }
-bodyContentText = text:bodyContentChar+ { return text.join(''); }
-bodyContentChar = escaped / !tabStop char:. { return char; }
 
-escaped = '\\' char:. { return char; }
-tabStop =  tabStopWithTransformation /  tabStopWithPlaceholder / tabStopWithoutPlaceholder / simpleTabStop
+bodyContent = content:(tabstop / choice / variable / text)* { return content; }
+
+innerBodyContent = content:(tabstop / choice / variable / nonCloseBraceText)* { return content; }
 
-simpleTabStop = '$' index:[0-9]+ {
-  return { index: parseInt(index.join("")), content: [] };
+tabstop = simpleTabstop / tabstopWithoutPlaceholder / tabstopWithPlaceholder / tabstopWithTransform
+
+simpleTabstop = '$' index:int {
+  return { index: makeInteger(index), content: [] }
 }
-tabStopWithoutPlaceholder = '${' index:[0-9]+ '}' {
-  return { index: parseInt(index.join("")), content: [] };
+
+tabstopWithoutPlaceholder = '${' index:int '}' {
+  return { index: makeInteger(index), content: [] }
 }
-tabStopWithPlaceholder = '${' index:[0-9]+ ':' content:placeholderContent '}' {
-  return { index: parseInt(index.join("")), content: content };
+
+tabstopWithPlaceholder = '${' index:int ':' content:innerBodyContent '}' {
+  return { index: makeInteger(index), content: content }
 }
-tabStopWithTransformation = '${' index:[0-9]+ substitution:transformationSubstitution '}' {
+
+tabstopWithTransform = '${' index:int substitution:transform '}' {
   return {
-    index: parseInt(index.join(""), 10),
+    index: makeInteger(index),
     content: [],
     substitution: substitution
-  };
+  }
 }
 
-placeholderContent = content:(tabStop / placeholderContentText / variable )* { return flatten(content); }
-placeholderContentText = text:placeholderContentChar+ { return coalesce(text); }
-placeholderContentChar = escaped / placeholderVariableReference / !tabStop !variable char:[^}] { return char; }
+choice = '${' index:int '|' choice:choicecontents '|}' {
+  return { index: makeInteger(index), choice: choice }
+}
 
-placeholderVariableReference = '$' digit:[0-9]+ {
-  return { index: parseInt(digit.join(""), 10), content: [] };
+choicecontents = elem:choicetext rest:(',' val:choicetext { return val } )* {
+  return [elem, ...rest]
 }
 
-variable = '${' variableContent '}' {
-  return ''; // we eat variables and do nothing with them for now
+choicetext = choicetext:(choiceEscaped / [^|,] / barred:('|' &[^}]) { return barred.join('') } )+ {
+  return choicetext.join('')
 }
-variableContent = content:(variable / variableContentText)* { return content; }
-variableContentText = text:variableContentChar+ { return text.join(''); }
-variableContentChar = !variable char:('\\}' / [^}]) { return char; }
 
-escapedForwardSlash = pair:'\\/' { return pair; }
+// Transform is applied when tabbed off
```

Comment:
Is this comment accurate? Looks like the transformation code is largely unchanged, and current behavior is to apply the transform immediately and upon every change to the input text.

Reply:
VS Code applies when tabbed off. But I guess leaving the existing behaviour also works.

I think I left that comment to remind me that whatever you're typing while filling it out doesn't matter, only what you leave it as.